packages/astro/src/core/errors/utils.ts

```diff
@@ -0,0 +1,303 @@
+import * as fs from 'fs';
+import stripAnsi from 'strip-ansi';
+import { fileURLToPath } from 'url';
+import { createLogger, ErrorPayload, Logger, LogLevel, ViteDevServer } from 'vite';
+import { SSRError } from '../../@types/astro.js';
+import { AstroErrorCodes } from './codes.js';
+import { AggregateError, AstroError, ErrorWithMetadata } from './errors.js';
+import { codeFrame } from './printer.js';
+
+/**
+ * Generate a payload for Vite's error overlay
+ */
+export function getViteErrorPayload(err: ErrorWithMetadata): ErrorPayload {
+	let plugin = err.plugin;
+	if (!plugin && err.hint) {
+		plugin = 'astro';
+	}
+	const message = `${err.message}\n\n${err.hint ?? ''}`;
+	// Vite doesn't handle tabs correctly in its frames, so let's replace them with spaces
+	const frame = err.frame?.replace(/\t/g, ' ');
+	return {
+		type: 'error',
+		err: {
+			...err,
+			frame: frame,
+			loc: {
+				file: err.loc?.file,
+				// If we don't have a line and column, Vite won't make a clickable link, so let's fake 0:0 if we don't have a location
+				line: err.loc?.line ?? 0,
+				column: err.loc?.column ?? 0,
+			},
+			plugin,
+			message: message.trim(),
+			stack: err.stack,
+		},
+	};
+}
+
+/**
+ * Get the line and character based on the offset
+ * @param offset The index of the position
+ * @param text The text for which the position should be retrived
+ */
+export function positionAt(
+	offset: number,
+	text: string
+): {
+	line: number;
+	column: number;
+} {
+	const lineOffsets = getLineOffsets(text);
+	offset = Math.max(0, Math.min(text.length, offset));
+
+	let low = 0;
+	let high = lineOffsets.length;
+	if (high === 0) {
+		return {
+			line: 0,
+			column: offset,
+		};
+	}
+
+	while (low <= high) {
+		const mid = Math.floor((low + high) / 2);
+		const lineOffset = lineOffsets[mid];
+
+		if (lineOffset === offset) {
+			return {
+				line: mid,
+				column: 0,
+			};
+		} else if (offset > lineOffset) {
+			low = mid + 1;
+		} else {
+			high = mid - 1;
+		}
+	}
+
+	// low is the least x for which the line offset is larger than the current offset
+	// or array.length if no line offset is larger than the current offset
+	const line = low - 1;
+	return { line, column: offset - lineOffsets[line] };
+}
+
+function getLineOffsets(text: string) {
+	const lineOffsets = [];
+	let isLineStart = true;
+
+	for (let i = 0; i < text.length; i++) {
+		if (isLineStart) {
+			lineOffsets.push(i);
+			isLineStart = false;
+		}
+		const ch = text.charAt(i);
+		isLineStart = ch === '\r' || ch === '\n';
+		if (ch === '\r' && i + 1 < text.length && text.charAt(i + 1) === '\n') {
+			i++;
+		}
+	}
+
+	if (isLineStart && text.length > 0) {
+		lineOffsets.push(text.length);
+	}
+
+	return lineOffsets;
+}
+
+export function enhanceViteSSRError(
+	error: Error,
+	filePath?: URL,
+	viteServer?: ViteDevServer
+): AstroError {
+	// Vite will give you better stacktraces, using sourcemaps.
+	if (viteServer) {
+		viteServer?.ssrFixStacktrace(error);
```

Comment:
This function some times throws, we were previously wrapping in a try/catch to account for this, see https://github.com/withastro/astro/pull/5206/files#diff-57bbfc11a4debc4ded58d61eedeb4a85e98644582748c7a9a9a16f1e24ee6a4dL51-L53

Reply:
This is when it was added https://github.com/withastro/astro/commit/0dba3b6f3fbd013f922fd11b9d6d977d165a512a